### ISSUE TITLE
Allow customizing `loc` for model validators

### ISSUE THREAD

**Issue by akx** [pydantic/pydantic#10668]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
  - (I have actually searched GH for similar requests and found #8092 and #9686, which touch upon this.) 
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Given something like

```python
from typing import Literal, Self

import pydantic


class ServiceConfiguration(pydantic.BaseModel):
    access_key: str | None = None
    endpoint_url: pydantic.HttpUrl | Literal[""] | None = None

    @pydantic.model_validator(mode="after")
    def check_access_key_format(self) -> Self:
        if self.endpoint_url:  # Customized endpoint, do not check format.
            return self
        if self.access_key and not self.access_key.islower():
            raise ValueError(
                "Access key must be lower case. The default service does not tolerate shouting and loud noises.")
        return self


ServiceConfiguration.model_validate({"access_key": "VERYVERYLOUD"})
```

I would like to be able to customize the `loc` for the resulting error to be `'access_key'`, since the error is actually an error in that field, so I can eventually format the `pydantic.ValidationError` to neatly nested JSON with a reference to the actual field that had the issue. 

Some ideas for implementing this:

* Allow setting `field_name` for `model_validator()`s too, and use it for `loc`. Not very dynamic, maybe overloads the `field_name` name too much?
* Read a `loc` or `_pydantic_loc` or similar field off the exception user code raises, use it if available. Simple enough, but also relies on the implicit magic that this is a thing, and creating such an exception is not idiomatic (and may not be possible, if the exception class has e.g. slots):
  ```python
  exc = ValueError("No shouting!")
  exc.loc = ("access_key",)
  raise exc
  ```
* A Pydantic-specific extra error info wrapper exception type?
  ```python
  raise pydantic.ErrorWithContext(ValueError("No shouting!"), loc=("access_key", ))
  ```
  Feels pretty elegant, should be easy to peel off and process in the Rust core code?
* Add `loc` to [`PydanticCustomError`](https://docs.pydantic.dev/dev/api/pydantic_core/#pydantic_core.PydanticCustomError)?

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by sydney-runkle** [comment]
Hi @akx,

Thanks for your inquiry.

You can do what you're hoping to with the following:

```py
from typing import Literal, Self

import pydantic
from pydantic_core import InitErrorDetails, PydanticCustomError
from pydantic import ValidationError


class ServiceConfiguration(pydantic.BaseModel):
    access_key: str | None = None
    endpoint_url: pydantic.HttpUrl | Literal[''] | None = None

    @pydantic.model_validator(mode='after')
    def check_access_key_format(self) -> Self:
        if self.endpoint_url:  # Customized endpoint, do not check format.
            return self
        if self.access_key and not self.access_key.islower():
            raise ValidationError.from_exception_data(
                'LowerCaseError',
                [
                    InitErrorDetails(
                        type=PydanticCustomError(
                            'str_not_lower',
                            'Access key must be lower case. The default service does not tolerate shouting and loud noises.',
                        ),
                        loc=('access_key',),
                        input=self.access_key,
                    ),
                ],
            )
        return self


try:
    ServiceConfiguration.model_validate({'access_key': 'VERYVERYLOUD'})
except ValidationError as e:
    print(e.errors())
    """
    [{'type': 'str_not_lower', 'loc': ('access_key',), 'msg': 'Access key must be lower case. The default service does not tolerate shouting and loud noises.', 'input': 'VERYVERYLOUD'}]
    """
```

I'll be expanding our docs on this subject shortly. Going to mark as resolved for now 👍 

Also note, #9686 has recently been closed (resolved), so that will take effect in v2.10!

**Comment by bblanchon** [comment]
@sydney-runkle, this solution is quite convoluted, especially for what seems to be a common use case.
Can't we have a shortcut like `raise ValidationError("No shouting!", loc=('access_key',))`?
